exonum/src/proto/schema/exonum/runtime.proto

```diff
@@ -66,34 +66,53 @@ message GenesisConfig {
   repeated InstanceInitParams builtin_instances = 3;
 }
 
-// The kind of ExecutionError.
+// The type of ExecutionError.
+// Note that ErrorKind doesn't communicate the source of the error, for this
+// see the 'runtime_id' and 'call_site' fields of ExecutionError.
 enum ErrorKind {
+  // An unexpected error which does not have a specific cause.
+  // The description of the error is the only source of information about this kind of errors.
   UNEXPECTED = 0;
+  // An error specific to the core. See core error codes for details.
   CORE = 1;
+  // An error specific to the certain runtime. See error codes of corresponding runtime for details.
   RUNTIME = 2;
+  // An error specific to the certain service. See error codes of corresponding service for details.
   SERVICE = 3;
+  // A common error which can occur in different contexts. See common error codes for details.
   COMMON = 4;
 }
 
 // Result of unsuccessful runtime execution.
+// ExecutionError message provides the information about the source of the error.
+// The source of error is determined as following:
+// - If both runtime ID and call site are set, then error originates in the service code.
+// - If runtime ID is set, and call site is not set, then error originates in the runtime code.
```

Comment:
The question not to this code, but Runtime spec. As only the runtime can know if an error occurred in service or its code, is it its responsibility to communicate that to the core?

@vitvakatu shall we modify anything?

Reply:
Setting call site is possible only in Core (as the method `set_call_site` is package-private). Though I don't understand how the Core will understand if an error occurred in service code.

Reply:
Core provides some guarantees on the requests that it makes to runtime:

If core requests a runtime to invoke a `BeforeTransactions` hook for service, then it's sure that this service exists in the runtime and runtime should be able to dispatch the call.
Thus, the only possible variant of managed error here is the error in the service code.

Error in runtime code there may happen only because of the bug in either runtime or core code, and for this purpose we have `Unexpected` error type, and even in this condition call site friendly will tell us that error happened when core requested runtime X to perform something for service Y, and the error description will contain something like "BUG: Core unexpectedly requested to invoke hook for non-existent service".

OTOH, if core performs a request to the runtime *itself* (rather than to a service), e.g. artifact deployment, there is obviously no service call site, and the source of (expected and valid) error will surely be in the runtime code.

Reply:
I see, thank you, @popzxc 

@dmitry-timofeev no changes are needed on our side. The only thing we can alter is to utilize `CommonError`, but if I understand it correctly, it is not entirely necessary.

Reply:
> Error in runtime code there may happen only because of the bug in either runtime or core code

Not necessarily. "Transaction with the given id does not exist", "Malformed arguments" — those errors happen in _runtime_ code during tx processing, which performs routing and decoding correspondingly. They may be caused by a client bug (invalid arguments are passed), a runtime bug (we can't route/decode correctly), or a service bug (it uses incorrect tx id).

Reply:
Hm.  As I see it:

If the destination service exists, then both of non-existent tx id and malformed arguments errors belong to the service code.

Runtime was requested just to pass the certain arguments to the service. Service is present, but it cannot process the request. Thus, the error originates in the service.

Even if the runtime is capable of deserializing arguments for the service call, it's still an additional feature that runtime authors did introduce to make service authors life easier. Service could attempt to deserialize data by itself and face the same issue.

One more point here is that `ExecutionError` in most cases denotes some sort of *expected* error, which can happen in a normal workflow. For bugs we have `Unexpected` error type, and IMHO we should not try to mix bugs with expected errors (caused by incorrect input provided by user), since hopefully `Unexpected` errors will never occur in production code.

Reply:
I don't agree that such errors belong to or originate in the service _code_ because they happen outside of it, they are raised by the runtime code. They are certainly related to a particular service, but do not originate in its code. 

The preserved id of the related service is definitely useful for debuggability. But the current doc says "originates in the service code". I suggest to relax it to "relates to a particular service", because the current phrasing says it is thrown from the service code.

Reply:
Sure, makes sense.